virtcontainers/proxy.go

```diff
@@ -161,3 +171,91 @@ type proxy interface {
 	//check if the proxy has watched the vm console.
 	consoleWatched() bool
 }
+
+func (p *proxyBuiltin) watchConsole(proto, console string, logger *logrus.Entry) (err error) {
+	var (
+		scanner *bufio.Scanner
+		conn    net.Conn
+	)
+
+	switch proto {
+	case consoleProtoUnix:
+		conn, err = net.Dial("unix", console)
+		if err != nil {
+			return err
+		}
+		// TODO: please see
+		// https://github.com/kata-containers/runtime/issues/1940.
+	case consoleProtoPty:
+		fallthrough
+	default:
+		return fmt.Errorf("unknown console proto %s", proto)
+	}
+
+	p.conn = conn
+
+	go func() {
+		scanner = bufio.NewScanner(conn)
+		for scanner.Scan() {
+			logger.WithFields(logrus.Fields{
+				"sandbox":   p.sandboxID,
+				"vmconsole": scanner.Text(),
+			}).Debug("reading guest console")
+		}
+
+		if err := scanner.Err(); err != nil {
+			if err == io.EOF {
+				logger.Info("console watcher quits")
+			} else {
+				logger.WithError(err).WithFields(logrus.Fields{
+					"console-protocol": proto,
+					"console-socket":   console,
+				}).Error("Failed to read agent logs")
+			}
+		}
+	}()
+
+	return nil
+}
+
+// check if the proxy has watched the vm console.
+func (p *proxyBuiltin) consoleWatched() bool {
+	return p.conn != nil
+}
+
+// start is the proxy start implementation for builtin proxy.
+// It starts the console watcher for the guest.
+// It returns agentURL to let agent connect directly.
+func (p *proxyBuiltin) start(params proxyParams) (int, string, error) {
+	if params.logger == nil {
+		return -1, "", fmt.Errorf("Invalid proxy parameter: proxy logger is not set")
+	}
+
+	if p.consoleWatched() {
+		return -1, "", fmt.Errorf("kata proxy running for sandbox %s", params.id)
+	}
+
+	params.logger.Debug("Starting kata proxy")
```

Comment:
@lifupan This will be logged for the the noProxy.start() as well which is not desired.
Similarly I would suggest changing the above Error message as well to something like "Error setting up console" for the noProxy case, else the user may think that an actual proxy is running looking at the logs/errors.

Reply:
Hi @amshinde, Updated!